src/SquidMath.h

```diff
@@ -123,5 +123,40 @@ SetToNaturalSumOrMax(S &var, Args... args)
     return var;
 }
 
+template <typename P, typename T, typename U>
+Optional<P>
+IntegralProduct(P, T t, U u)
```

Comment:
This function feels more similar to IncreaseSum() rather than NaturalSum(). Perhaps it should be refactored as IncreaseProduct(), losing its first argument?! We may need to provide a high-level wrapper to avoid type guessing, but such refactoring might simplify things overall.